JS mini project

ahngh0113 opened the pull request:
JS mini project

validation function
    1. type check
    2. negative num check
    
logic function
    1. reduce
    2. filter
    3. find
    4. findIndex
    5. forEach
    6. map
    7. for
    8. while

ahngh0113 committed:
asset: add folder for js mini project

ahngh0113 committed:
feat: make function for goods recommend

validation function
1. type check
2. negative num check

logic function
1. reduce
2. filter
3. find
4. findIndex
5. forEach
6. map
7. for
8. while

heejinlee07 commented on boilerplate/index.html:
```diff
@@ -0,0 +1,10 @@
+<body>
+  <script src="main.js"></script>
+</body>
+</html>
```
엔터를 입력하여 줄바꿈 한 줄 넣어주세요

heejinlee07 commented on boilerplate/main.js:
```diff
@@ -0,0 +1,240 @@
+
+// 사용자 입력 받기
+const line = prompt("최대 금액을 입력해주세요.");
+const amount = +line;
```
prompt()의 결과가 문자열이라서 숫자로 변환하기 위해 `+`기호를 작성해주신 것 같습니다. 
일반적으로 코드 상에서 문자열을 숫자로 변환할 때는 `+`기호를 이용해서 강제 변환하는 것보다는 
javaScript 내장 함수를 이용하는 것이 좋습니다. 실무에서도 보통 아래 함수를 이용해서 
문자열을 숫자로 변환합니다.

- parseInt(line);
- Number(line);

보다 자세한 내용은 아래 링크를 참조하세요:)
https://www.delftstack.com/ko/howto/javascript/convert-string-to-number-javascript/

heejinlee07 commented on boilerplate/main.js:
```diff
@@ -0,0 +1,240 @@
+const amount = +line;
+
+// 주어진 금액으로 살 수 있는 가장 비싼 상품을 구함
+const item = getItemByAmount(data, amount);
```
data가 위에 const로 선언되어 있기 때문에 인수로 전달하지 않아도 됩니다.

heejinlee07 commented on boilerplate/main.js:
```diff
@@ -0,0 +1,240 @@
+// 사용자의 입력한 값의 타입이 숫자인지 확인하는 함수
+// 숫자이면 false, 숫자가 아니면 true를 반환
+function typeValid(amount) {
+  return isNaN(amount);
```
isNaN을 사용할 때는 주의해서 사용해야 합니다.
만약 아래와 같이 null값이 들어왔다면 결과가 true일 것 같지만 false입니다.
`isNaN(null)` // false

입력된 값이 숫자인지 아닌지를 판별할 때는 `typeof`와 `===`연산자를 사용하여 아래와 같이 
사용하는 것이 더 정확합니다.

```
const count = 200;

if(typeof count === 'number') console.log('Number');
if(typeof count !== 'number') console.log('Not A Number');
```

참고로 `==`과 `===`은 아래와 같은 차이점이 있습니다. 실무에서는 거의 `===`을 사용합니다. 
`===`이 더욱 엄격한 비교 연산자이기 때문입니다. 
- == 동등 비교: x == y x와 y의 값이 같음
- === 일치 비교: x === y x와 y의 값과 타입이 같음



자세한 내용은 아래 링크를 참조하세요.
https://mygumi.tistory.com/335

heejinlee07 commented on boilerplate/main.js:
```diff
@@ -0,0 +1,240 @@
+
+// 사용자의 입력한 값이 0 이상의 숫자인지 확인하는 함수
+// 0 이상이면 false, 0 보다 작으면 true를 반환
+function negValid(amount) {
```
negValid가 무슨의미인가요? 함수나 변수명은 이름을 통해 동작을 가늠해볼 수 있도록 명료하게 작성하는 것이 좋습니다.

heejinlee07 reviewed (commented):
LGTM😃
고생많으셨습니다!

여러 가지 방법을 통해 로직을 구성해주셨네요.
로직을 처리하기 위해 여러가지 방법을 사용해보신 점이 좋습니다.
특히 고차 배열메서드를 쓰셔서 해결한 점이 좋습니다.

다만 pr리뷰 올리실 때 주석처리한 코드는 남기지 않으시는 것이 좋습니다.
코드에 대한 설명 외의 주석은 pr시 남아있지 않아야 합니다.

ahngh0113 committed:
feat: line break

ahngh0113 committed:
feat: remove not necessary annotation and change function logic

remove not necessary annotation
change + to Number()
change isNaN() to amount !== NaN
